### DISCUSSION TITLE
choria on puppet 8

### DISCUSSION THREAD

**Discussion by dom0317** [choria-io/general#42] [General]
Does choria support / work with puppet 8?
Or is there anyone using it with puppet 8?

We are looking at replacing mcollective with choria and upgrading from puppet 3.8 to puppet 7 or 8. This will be on RHEL 8. We would like to go to puppet 8. On preliminary investigations choria is working with with puppet 7 but hitting an issue with puppet 8.

**Comment by ripienaar** [comment]
It should now work, I didnt announce the release yet on the blog as I want to give ample period for feedback but I think at this point if you follow the normal guide it will just work on puppet 8.

**Reply by dom0317** [reply]
Thanks, that is good news.

In limited testing I found it does work ok. I can successfully run some of our existing mcollective actions. 
I did find 3 small issues:

1) In  choria-mcorpc-support-2.26.2/lib/mcollective/shell.rb
at the end of runcommand method changed:
@status.thread.killd 
to
@status.thread.kill if respond_to?("@status.thread")
Otherwise got a "undefined method `thread' " error on mco rpc actions and mco plugin package. 

2) In choria-mcollective_agent_puppet-2.4.2/files/mcollective/agent/puppet.rb
in run method changed: 
env = ENV.clone 
to
env = ENV.to_h

3) choria-mcorpc-support-2.26.2/lib/mcollective/pluginpackager/forge_packager.rb
This only gave a warning message, but changed:
erb = ERB.new(File.read(infile), 0, "-")
to
erb = ERB.new(File.read(infile), trim_mode: "-")

**Reply by ripienaar** [reply]
Issues and PRs for these would be great

**Comment by smortex** [comment]
Hey!  I updated one $WORK site today to Puppet 8.

Two choria-related issues when using the latest modules from the forge prevent choria to work:
1. https://github.com/choria-plugins/mcollective_choria/pull/27
2. https://github.com/choria-plugins/puppet-agent/pull/61

After adjusting the files by hand, everything seems to be fine.

I did not hurt these issues on my home infra because:
1. there I run FreeBSD and is this case the module ensure the package of the ruby gem is installed rather than installing the gem directly;
2. FreeBSD currently default to Ruby 3.1 which has not the `ENV.clone` deprecation.

New releases of these plugins should make Puppet 8 support complete.

**Reply by smortex** [reply]
In complement, some modules still use the removed `File#exists?`, link to PR that fix them:

* https://github.com/choria-io/puppet-mcollective/pull/330
* https://github.com/choria-plugins/package-agent/pull/27 (fixed while petting CI)
* https://github.com/choria-plugins/filemgr-agent/pull/5

These also would require a new release to fix Puppet 8 support.

**Reply by ripienaar** [reply]
Thanks @smortex I've merged all these, will look to do releases this week as I think we're getting quite close now

**Reply by smortex** [reply]
Cool!

We also need a release of choria-mcorpc-support that include:
* https://github.com/choria-io/mcorpc-ruby-support/pull/187
